core/src/databases/table_schema.rs

```diff
@@ -0,0 +1,219 @@
+use std::collections::HashMap;
+
+use anyhow::{anyhow, Result};
+
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+#[serde(rename_all = "lowercase")]
+pub enum TableSchemaFieldType {
+    Int,
+    Float,
+    Text,
+    Bool,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub struct TableSchema(HashMap<String, TableSchemaFieldType>);
+
+impl TableSchema {
+    pub fn from_rows(rows: Vec<Value>) -> Result<Self> {
+        let mut schema = HashMap::new();
+
+        for (row_index, row) in rows.iter().enumerate() {
+            let object = row
+                .as_object()
+                .ok_or_else(|| anyhow!("Row {} is not an object", row_index))?;
+
+            for (k, v) in object {
+                if v.is_null() {
+                    continue;
+                }
+
+                let value_type = match v {
+                    Value::Bool(_) => TableSchemaFieldType::Bool,
+                    Value::Number(x) => {
+                        if x.is_i64() {
+                            TableSchemaFieldType::Int
+                        } else {
+                            TableSchemaFieldType::Float
+                        }
+                    }
+                    Value::String(_) | Value::Object(_) | Value::Array(_) => {
+                        TableSchemaFieldType::Text
+                    }
+                    _ => unreachable!(),
+                };
+
+                if let Some(existing_type) = schema.get(k) {
+                    if existing_type != &value_type {
+                        return Err(anyhow!(
```

Comment:
here attention is needed because it will happen that in some legitimate text fields, the text value is "true" or "2332" so we will get errs on valid tables regularly
OTOH we do not want to treat everything as a text field, if we first check the value is ok as text

A possible solution: iterate on keys rather than rows, for each key match values with most narrow types first as you did above; then => if all match the narrowest type, cool, if all but 1 match (or 90% or whatever) => err, likely malformed data, if less than 90% match => probably a text field

Totally worth discussing since tradeoffy

Reply:
> here attention is needed because it will happen that in some legitimate text fields, the text value is "true" or "2332" so we will get errs on valid tables regularly

Not sure I fully understand this -- Boolean and String are distinct types in JSON. `"true"` is a string while `true` is a bool. They shouldn't clash with eachother IMO ?

Otherwise, if we do have an issue, we could indeed implement "widening" the type as string, but not convinced we'll need it

Reply:
Basically my point is that our schema should ideally be a function of the JSON types and not of the content of fields. Do you think we'll have many cases where number / booleans are represented as strings ?

Reply:
As discussed IRL, my point was e.g. regarding CSVs which may not have type information
=> as you pointed out, the check is better done at ingestion time so here we have a guaranteed match between JSON schema and SQL schema